<!DOCTYPE html><html lang="en" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0,viewport-fit=cover"><title>Hexo | Hexo</title><meta name="author" content="John Doe"><meta name="copyright" content="John Doe"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="7.1. 索引到目前为止，我们已经从数据集中删除了低质量细胞、环境RNA污染和双倍体，数据以计数矩阵的形式作为形状为细胞数×基因数的数字矩阵可用。这些计数表示单细胞RNA测序实验中分子的捕获、反转录和测序。这些步骤中的每一步都会为相同的细胞测量计数深度添加一定的变异度，因此计数数据中细胞间基因表达的差异可能仅仅是由于采样效应造成的。这意味着数据集和因此计数矩阵仍然包含广泛变化的方差项。分析数据集通">
<meta property="og:type" content="article">
<meta property="og:title" content="Hexo">
<meta property="og:url" content="http://example.com/2023/04/20/7.Normalization/index.html">
<meta property="og:site_name" content="Hexo">
<meta property="og:description" content="7.1. 索引到目前为止，我们已经从数据集中删除了低质量细胞、环境RNA污染和双倍体，数据以计数矩阵的形式作为形状为细胞数×基因数的数字矩阵可用。这些计数表示单细胞RNA测序实验中分子的捕获、反转录和测序。这些步骤中的每一步都会为相同的细胞测量计数深度添加一定的变异度，因此计数数据中细胞间基因表达的差异可能仅仅是由于采样效应造成的。这意味着数据集和因此计数矩阵仍然包含广泛变化的方差项。分析数据集通">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="https://i.loli.net/2021/02/24/5O1day2nriDzjSu.png">
<meta property="article:published_time" content="2023-04-20T14:28:51.386Z">
<meta property="article:modified_time" content="2023-04-20T14:30:47.094Z">
<meta property="article:author" content="John Doe">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://i.loli.net/2021/02/24/5O1day2nriDzjSu.png"><link rel="shortcut icon" href="/img/favicon.png"><link rel="canonical" href="http://example.com/2023/04/20/7.Normalization/index.html"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox/fancybox.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = { 
  root: '/',
  algolia: undefined,
  localSearch: undefined,
  translate: undefined,
  noticeOutdate: undefined,
  highlight: {"plugin":"highlighjs","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":false},
  copy: {
    success: 'Copy successfully',
    error: 'Copy error',
    noSupport: 'The browser does not support'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  dateSuffix: {
    just: 'Just',
    min: 'minutes ago',
    hour: 'hours ago',
    day: 'days ago',
    month: 'months ago'
  },
  copyright: undefined,
  lightbox: 'fancybox',
  Snackbar: undefined,
  source: {
    justifiedGallery: {
      js: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.js',
      css: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: false,
  isAnchor: false,
  percent: {
    toc: true,
    rightside: false,
  }
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: 'Hexo',
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2023-04-20 22:30:47'
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
    win.saveToLocal = {
      set: function setWithExpiry(key, value, ttl) {
        if (ttl === 0) return
        const now = new Date()
        const expiryDay = ttl * 86400000
        const item = {
          value: value,
          expiry: now.getTime() + expiryDay,
        }
        localStorage.setItem(key, JSON.stringify(item))
      },

      get: function getWithExpiry(key) {
        const itemStr = localStorage.getItem(key)

        if (!itemStr) {
          return undefined
        }
        const item = JSON.parse(itemStr)
        const now = new Date()

        if (now.getTime() > item.expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return item.value
      }
    }
  
    win.getScript = url => new Promise((resolve, reject) => {
      const script = document.createElement('script')
      script.src = url
      script.async = true
      script.onerror = reject
      script.onload = script.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        script.onload = script.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(script)
    })
  
    win.getCSS = (url,id = false) => new Promise((resolve, reject) => {
      const link = document.createElement('link')
      link.rel = 'stylesheet'
      link.href = url
      if (id) link.id = id
      link.onerror = reject
      link.onload = link.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        link.onload = link.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(link)
    })
  
      win.activateDarkMode = function () {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = function () {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
          if (t === 'dark') activateDarkMode()
          else if (t === 'light') activateLightMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    
    const detectApple = () => {
      if(/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)){
        document.documentElement.classList.add('apple')
      }
    }
    detectApple()
    })(window)</script><meta name="generator" content="Hexo 6.3.0"></head><body><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img is-center"><img src="https://i.loli.net/2021/02/24/5O1day2nriDzjSu.png" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"/></div><div class="sidebar-site-data site-data is-center"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">5</div></a><a href="/tags/"><div class="headline">Tags</div><div class="length-num">0</div></a><a href="/categories/"><div class="headline">Categories</div><div class="length-num">0</div></a></div><hr/></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header"><nav id="nav"><span id="blog-info"><a href="/" title="Hexo"><span class="site-name">Hexo</span></a></span><div id="menus"><div id="toggle-menu"><a class="site-page" href="javascript:void(0);"><i class="fas fa-bars fa-fw"></i></a></div></div></nav><div id="post-info"><h1 class="post-title">No title</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">Created</span><time class="post-meta-date-created" datetime="2023-04-20T14:28:51.386Z" title="Created 2023-04-20 22:28:51">2023-04-20</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">Updated</span><time class="post-meta-date-updated" datetime="2023-04-20T14:30:47.094Z" title="Updated 2023-04-20 22:30:47">2023-04-20</time></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title=""><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">Post View:</span><span id="busuanzi_value_page_pv"><i class="fa-solid fa-spinner fa-spin"></i></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><h2 id="7-1-索引"><a href="#7-1-索引" class="headerlink" title="7.1. 索引"></a>7.1. 索引</h2><p>到目前为止，我们已经从数据集中删除了低质量细胞、环境RNA污染和双倍体，数据以计数矩阵的形式作为形状为细胞数×基因数的数字矩阵可用。这些计数表示单细胞RNA测序实验中分子的捕获、反转录和测序。这些步骤中的每一步都会为相同的细胞测量计数深度添加一定的变异度，因此计数数据中细胞间基因表达的差异可能仅仅是由于采样效应造成的。这意味着数据集和因此计数矩阵仍然包含广泛变化的方差项。分析数据集通常是具有挑战性的，因为许多统计方法都假设具有均匀方差结构的数据。</p>
<blockquote>
<p>伽马-泊松分布<br>UMI数据的理论和实证建立的模型是伽马泊松分布，它意味着平方均方差关系，$\operatorname{Var}[Y]&#x3D;\mu+\alpha \mu^{2}$<br>，平均值为 $μ$和过度离群为$α$。对于$α&#x3D;0$，这是Poisson分布，描述了超过Poisson分布的额外方差。”</p>
</blockquote>
<p>“归一化”数据预处理步骤的目的是通过将可观测方差缩放到指定范围，对数据集中的原始计数进行调整以处理变量采样效应。实践中使用多种不同的归一化技术，复杂程度也各不相同。它们大多设计成这样的方式，以便后续分析任务和它们的基础统计方法可以应用。</p>
<p>Ahlmann-Eltze和Huber在2022年发表的最近一项基准测试[Ahlmann-Eltze and Huber, 2022]比较了22种单细胞数据转换方法的效果。该基准测试基于细胞图形与基本事实的重叠情况比较了不同归一化技术的表现。我们想强调的是，仍未有一份完整的基准测试，该测试还比较了归一化对各种不同的下游分析任务的影响。我们建议分析师仔细选择归一化方法，并始终依赖于后续分析任务。</p>
<p>本章将向读者介绍三种不同的归一化技术，即移位对数转换、scran归一化和Pearson残差的分析逼近。移位对数有助于稳定后续的维度缩减和鉴定差异表达基因。Scran已经过广泛测试，并用于批量校正任务，分析Pearson残差尤其适用于选择具有生物变异性的基因和识别稀有细胞类型。</p>
<p>我们首先导入所有必需的Python软件包，并加载我们过滤了低质量细胞、移除了环境RNA并评分为双细胞的数据集。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> scanpy <span class="keyword">as</span> sc</span><br><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"><span class="keyword">import</span> seaborn <span class="keyword">as</span> sns</span><br><span class="line"><span class="keyword">from</span> matplotlib <span class="keyword">import</span> pyplot <span class="keyword">as</span> plt</span><br><span class="line"><span class="keyword">import</span> anndata2ri</span><br><span class="line"><span class="keyword">import</span> logging</span><br><span class="line"><span class="keyword">from</span> scipy.sparse <span class="keyword">import</span> issparse</span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> rpy2.rinterface_lib.callbacks <span class="keyword">as</span> rcb</span><br><span class="line"><span class="keyword">import</span> rpy2.robjects <span class="keyword">as</span> ro</span><br><span class="line"></span><br><span class="line">sc.settings.verbosity = <span class="number">0</span></span><br><span class="line">sc.settings.set_figure_params(</span><br><span class="line">    dpi=<span class="number">80</span>,</span><br><span class="line">    facecolor=<span class="string">&quot;white&quot;</span>,</span><br><span class="line">    <span class="comment"># color_map=&quot;YlGnBu&quot;,</span></span><br><span class="line">    frameon=<span class="literal">False</span>,</span><br><span class="line">)</span><br><span class="line"></span><br><span class="line">rcb.logger.setLevel(logging.ERROR)</span><br><span class="line">ro.pandas2ri.activate()</span><br><span class="line">anndata2ri.activate()</span><br><span class="line"></span><br><span class="line">%load_ext rpy2.ipython</span><br></pre></td></tr></table></figure>
<blockquote>
<ul>
<li>sc.settings.verbosity &#x3D; 0：设置了输出信息的详细程度为0，即只输出重要信息，不显示调试信息。</li>
<li>sc.settings.set_figure_params()：设置了绘图参数，包括dpi分辨率、背景颜色、图形边框等。</li>
<li>rcb.logger.setLevel(logging.ERROR)：设置了R的日志输出级别，只输出错误级别信息。</li>
<li>ro.pandas2ri.activate()：激活了Pandas和R之间的转换接口。</li>
<li>anndata2ri.activate()：激活了Anndata和R之间的转换接口。</li>
<li>%load_ext rpy2.ipython：加载了R语言的IPython内核。</li>
</ul>
</blockquote>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">adata = sc.read(</span><br><span class="line">    filename=<span class="string">&quot;s4d8_quality_control.h5ad&quot;</span>,</span><br><span class="line">    backup_url=<span class="string">&quot;https://figshare.com/ndownloader/files/40014331&quot;</span>,</span><br><span class="line">)</span><br></pre></td></tr></table></figure>
<blockquote>
<p>加载数据集。如果下载不了 可以后台私信我。</p>
</blockquote>
<p>我们现在可以检查在质量控制过程中已经计算过的原始计数的分布。在标准的单细胞分析管道中，可以省略这一步骤，但是了解不同的标准化概念可能会有所帮助。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">p1 = sns.histplot(adata.obs[<span class="string">&quot;total_counts&quot;</span>], bins=<span class="number">100</span>, kde=<span class="literal">False</span>)</span><br></pre></td></tr></table></figure>
<blockquote>
<p>adata.obs[“total_counts”]这个变量的直方图，将直方图分成了100个小区间（即bins&#x3D;100），不采用核密度估计（即kde&#x3D;False）。最终的绘图结果存储在变量p1中，可以通过调用p1来查看或保存绘图结果。</p>
</blockquote>
<p><img src="https://typorapicz.oss-cn-hangzhou.aliyuncs.com/img/1681884438932-686a6e7a-a1c4-4f2b-8518-f44c1c34f580.png" alt="image.png"><br>7.2. 移位对数#<br>我们将介绍的第一种规范化技术是移位对数，它基于三角法 [Dorfman，1938]。 三角法将非线性函数<br>$f(Y)$ 应用于原始计数 $Y$并旨在使数据集中的方差更加相似。<br>通过移位对数来解决这个问题。<br>$f(y)&#x3D;\log \left(\frac{y}{s}+y_{0}\right)$<br>以$y$为原始计数，以$s$为所谓的大小因子，以$y0$描述假计数。大小因子为每个细胞确定，以考虑采样效应和不同细胞大小的变化。<br>细胞$c$的大小因子可以通过计算得出。<br> $s_{c}&#x3D;\frac{\sum_{c} y_{g c}}{L}$</p>
<p>对于一个基因$g$和$L$描述目标总和的基因组，有不同的方法来确定数据中的大小因子。我们将在本节中利用Scanpy的默认值，其中$L$是数据集中原始计数深度的中位数。许多分析模板使用固定值$L$，例如$L&#x3D;10^{5}, \text { or } L&#x3D;10^{6}$，导致通常称为每百万计数（CPM）的值。对于初学者来说，这些值可能看似任意，但它可以导致比单细胞数据集中通常看到的更大的过度分散。</p>
<blockquote>
<p>过度离散（<strong>Overdispersion</strong>）<br>过度离散度是指数据集中存在比预期更大的变异性。</p>
</blockquote>
<p>“移位对数”是一种快速归一化技术，优于其他方法，可揭示数据集的潜在结构（特别是在主成分分析之后）并有助于稳定方差，以进行后续降维和识别差异表达基因。现在我们将检查如何将此归一化方法应用于我们的数据集。通过在<code>target_sum = None</code>的情况下运行<code>pp.normalized_total</code>，可以方便地使用scanpy调用移位对数。由于我们希望在本教程中探索三种不同的归一化技术，因此将<code>inplace</code>参数设置为<code>False</code>。现在，第二步使用缩放计数，并且我们获得了第一次归一化计数矩阵。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">scales_counts = sc.pp.normalize_total(adata, target_sum=<span class="literal">None</span>, inplace=<span class="literal">False</span>)</span><br><span class="line"><span class="comment"># log1p transform</span></span><br><span class="line">adata.layers[<span class="string">&quot;log1p_norm&quot;</span>] = sc.pp.log1p(scales_counts[<span class="string">&quot;X&quot;</span>], copy=<span class="literal">True</span>)</span><br></pre></td></tr></table></figure>
<blockquote>
<p>这段代码的作用是对adata数据进行归一化和log1p转换处理。<br>首先，sc.pp.normalize_total()函数将adata中的每个细胞上的每个基因在所有细胞共现的基因数量的比例进行归一化，可以指定目标和也可以不指定，然后返回一个经过归一化的大小比例数据。这个数据被赋值给了scales_counts。<br>接下来，sc.pp.log1p()函数对scales_counts[“X”]进行单调递增的log1p转换，这个转换可以压缩值域，将数值偏高的基因表达值进行压缩，使得整个数据符合更好的正态分布。结果被保存在adata.layers[“log1p_norm”]中。 若要得到这变换后的数据，可以通过adata[:, :] 调用。<br>如果inplace&#x3D;True，则可以将这些变换应用到原始数据上，而不是像这样在副本上进行。</p>
</blockquote>
<p>我们现在可以检查我们的计数分布在应用移位对数后如何改变，并将其与我们的原始（但经过过滤的）数据集中的总计数进行比较。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">fig, axes = plt.subplots(<span class="number">1</span>, <span class="number">2</span>, figsize=(<span class="number">10</span>, <span class="number">5</span>))</span><br><span class="line">p1 = sns.histplot(adata.obs[<span class="string">&quot;total_counts&quot;</span>], bins=<span class="number">100</span>, kde=<span class="literal">False</span>, ax=axes[<span class="number">0</span>])</span><br><span class="line">axes[<span class="number">0</span>].set_title(<span class="string">&quot;Total counts&quot;</span>)</span><br><span class="line">p2 = sns.histplot(adata.layers[<span class="string">&quot;log1p_norm&quot;</span>].<span class="built_in">sum</span>(<span class="number">1</span>), bins=<span class="number">100</span>, kde=<span class="literal">False</span>, ax=axes[<span class="number">1</span>])</span><br><span class="line">axes[<span class="number">1</span>].set_title(<span class="string">&quot;Shifted logarithm&quot;</span>)</span><br><span class="line">plt.show()</span><br></pre></td></tr></table></figure>
<blockquote>
<p>该代码使用Python中的matplotlib和seaborn库绘制两个直方图子图，其中第一个子图展示数据集的总计数（total_counts），第二个子图展示经过数据调整后的对数值（log1p_norm）。以下是代码的解释：</p>
<ol>
<li><code>fig, axes = plt.subplots(1, 2, figsize=(10, 5))</code>: 这一行代码创建了一个包含两个子图的图片，并将其保存在fig变量中。同时，创建了一个变量axes来存放每一个子图的坐标轴。 </li>
<li><code>p1 = sns.histplot(adata.obs[&quot;total_counts&quot;], bins=100, kde=False, ax=axes[0])</code>: 这一行代码使用Seaborn库中的<code>histplot</code>函数来绘制第一个子图。在这里，数据集的总计数存放在<code>adata.obs[&quot;total_counts&quot;]</code>中，并将其分成100个bin。通过将<code>ax</code>参数设置为<code>axes[0]</code>，将绘制第一个子图。 </li>
<li><code>axes[0].set_title(&quot;Total counts&quot;)</code>: 该代码行为第一个子图设置了一个标题：总计数。 </li>
<li><code>p2 = sns.histplot(adata.layers[&quot;log1p_norm&quot;].sum(1), bins=100, kde=False, ax=axes[1])</code>: 与第2步相似，这一行代码使用<code>histplot</code>函数绘制第二个子图。在这里，<code>adata.layers[&quot;log1p_norm&quot;].sum(1)</code>是根据列(axis&#x3D;1)的总和来计算出数据调整后的对数值。同样，通过将ax参数设置为axes[1]，第二个子图被绘制在第一个子图的右侧。 </li>
<li><code>axes[1].set_title(&quot;Shifted logarithm&quot;)</code>: 该行代码为第二个子图设置了标题：调整后的对数值。 </li>
<li><code>plt.show()</code>: 这行代码将创建的图片展示出来。</li>
</ol>
</blockquote>
<p><img src="https://typorapicz.oss-cn-hangzhou.aliyuncs.com/img/1681885171925-4720a674-750c-4fce-abbd-9e97de85a8c5.png" alt="image.png"><br>第二种基于delta方法的归一化方法是Scran的基于池化的大小因子估计方法。Scran遵循与移位对数相同的原则，通过计算$f(y)&#x3D;\log \left(\frac{y}{s}+y_{0}\right)$描述一个伪计数。现在唯一的区别是，Scran利用反卷积方法，基于细胞池的基因进行线性回归来估计大小因子。该方法旨在更好地考虑数据集中所有细胞的计数深度差异。</p>
<p>细胞被分割成池，并且Scran使用基于基因的线性回归估计池基尺度因子。Scran已经广泛用于批次校正任务，并且可以很容易地使用相应的R包来调用。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">from</span> scipy.sparse <span class="keyword">import</span> csr_matrix, issparse</span><br></pre></td></tr></table></figure>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">%%R</span><br><span class="line">library(scran)</span><br><span class="line">library(BiocParallel)</span><br></pre></td></tr></table></figure>
<p>scran 需要一个粗糙聚类输入来提高规模因子估计性能。在本教程中，我们使用简单的预处理方法，并以低分辨率对数据进行聚类，以获得输入以进行规模因子估算。基本预处理包括假设所有规模因子相等（库大小归一化为百万计数 - CPM），并对计数数据进行对数变换。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Preliminary clustering for differentiated normalisation</span></span><br><span class="line">adata_pp = adata.copy()</span><br><span class="line">sc.pp.normalize_total(adata_pp)</span><br><span class="line">sc.pp.log1p(adata_pp)</span><br><span class="line">sc.pp.pca(adata_pp, n_comps=<span class="number">15</span>)</span><br><span class="line">sc.pp.neighbors(adata_pp)</span><br><span class="line">sc.tl.leiden(adata_pp, key_added=<span class="string">&quot;groups&quot;</span>)</span><br></pre></td></tr></table></figure>
<blockquote>
<p>这段代码是用于数据分析中的基本聚类过程。它对数据进行了不同的归一化处理，并使用PCA方法对数据进行降维处理。随后，使用邻域图构建近邻图，并使用Leiden算法进行聚类，以便将数据划分为不同的组。<br>具体来说，代码的运行过程如下：</p>
<ul>
<li><code>adata_pp = adata.copy()</code>原始数据被复制到新变量adata_pp中。</li>
<li><code>sc.pp.normalize_total(adata_pp)</code>使用normalize_total函数对数据进行总体归一化，以便在后续的处理中将不同的细胞进行比较。</li>
<li><code>sc.pp.log1p(adata_pp)</code>使用log1p函数对数据进行对数变换，以便使数据的分布更加接近正态分布。</li>
<li><code>sc.pp.pca(adata_pp, n_comps=15)</code>使用PCA方法对数据进行降维处理，并将n_comps参数设置为15，以便保留15个主成分。</li>
<li><code>sc.pp.neighbors(adata_pp)</code>使用邻域图构建近邻图，以便在下一步聚类时考虑细胞之间的相似度。</li>
<li><code>sc.tl.leiden(adata_pp, key_added=&quot;groups&quot;)</code>使用Leiden算法对数据进行聚类，并将聚类结果存储在key_added参数所指定的变量中。</li>
</ul>
</blockquote>
<p>通过这些预处理步骤，原始数据可以更好地适应后续的聚类算法，从而在分类和识别细胞类型时发挥更好的作用。</p>
<p>我们现在将 data_mat 和我们计算的组添加到我们的 R 环境中。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">data_mat = adata_pp.X.T</span><br><span class="line"><span class="comment"># convert to CSC if possible. See https://github.com/MarioniLab/scran/issues/70</span></span><br><span class="line"><span class="keyword">if</span> issparse(data_mat):</span><br><span class="line">    <span class="keyword">if</span> data_mat.nnz &gt; <span class="number">2</span>**<span class="number">31</span> - <span class="number">1</span>:</span><br><span class="line">        data_mat = data_mat.tocoo()</span><br><span class="line">    <span class="keyword">else</span>:</span><br><span class="line">        data_mat = data_mat.tocsc()</span><br><span class="line">ro.globalenv[<span class="string">&quot;data_mat&quot;</span>] = data_mat</span><br><span class="line">ro.globalenv[<span class="string">&quot;input_groups&quot;</span>] = adata_pp.obs[<span class="string">&quot;groups&quot;</span>]</span><br></pre></td></tr></table></figure>
<blockquote>
<ol>
<li>将adata_pp对象的数据矩阵转置，得到转置后的矩阵data_mat；</li>
<li>判断data_mat是否为稀疏矩阵，若是稀疏矩阵，则进行转换，并将其转换为COO或CSC格式；</li>
<li>将Python环境中的data_mat矩阵通过R语言的robjects包转换为R语言中的全局变量data_mat；</li>
<li>将adata_pp对象中的obs数据框的“groups”列赋值给R语言中的全局变量input_groups。</li>
</ol>
</blockquote>
<p>现在我们也可以删除 anndata 对象的副本了，因为我们已经获得了运行 scran 所需的所有对象。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">del</span> adata_pp</span><br></pre></td></tr></table></figure>
<p>我们现在根据之前计算的细胞组计算大小因子。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">%%R -o size_factors</span><br><span class="line"></span><br><span class="line">size_factors = sizeFactors(</span><br><span class="line">    computeSumFactors(</span><br><span class="line">        SingleCellExperiment(</span><br><span class="line">            <span class="built_in">list</span>(counts=data_mat)), </span><br><span class="line">            clusters = input_groups,</span><br><span class="line">            <span class="built_in">min</span>.mean = <span class="number">0.1</span>,</span><br><span class="line">            BPPARAM = MulticoreParam()</span><br><span class="line">    )</span><br><span class="line">)</span><br></pre></td></tr></table></figure>
<blockquote>
<p>这段代码是利用R语言进行转录组数据的标准化处理，特别是single-cell RNA-seq（单细胞RNA序列）的数据处理。主要功能是计算大小因子（size factors），用于调整每个样本的reads（reads数）数量，从而保证不同样本之间的reads计数可以进行比较。下面是代码的解释：</p>
<ul>
<li>“%%R -o size_factors” 表示在jupyter notebook（一个交互式笔记本）中运行R语言并输出size_factors的结果。</li>
<li>“size_factors &#x3D; sizeFactors()” 定义了一个名为size_factors的R对象，它将存储所有计算结果。</li>
<li>“computeSumFactors()” 函数计算每个样本的规格化因子（normalization factors），它们根据每个样本中基因的总读数以及每个基因的可变性来确定。</li>
<li>“SingleCellExperiment()” 函数以数据框（matrix）的形式载入数据。</li>
<li>“clusters &#x3D; input_groups” 表示将数据分组，用于计算每个样本的规格化因子。</li>
<li>“min.mean &#x3D; 0.1” 表示过滤掉低reads数量的样本。</li>
<li>“BPPARM &#x3D; Multicore Param()” 是R语言实现并行计算的一种方式，以加快计算速度。</li>
</ul>
</blockquote>
<p>我们在<code>.obs</code>文件中保存了<code>size_factors</code>，现在可以对数据进行归一化处理，随后应用log1p变换。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">adata.obs[<span class="string">&quot;size_factors&quot;</span>] = size_factors</span><br><span class="line">scran = adata.X / adata.obs[<span class="string">&quot;size_factors&quot;</span>].values[:, <span class="literal">None</span>]</span><br><span class="line">adata.layers[<span class="string">&quot;scran_normalization&quot;</span>] = csr_matrix(sc.pp.log1p(scran))</span><br></pre></td></tr></table></figure>
<blockquote>
<p>第一步是为adata添加一个名为”size_factors”的观测值，值为变量”size_factors”。第二步是进行一个简单的数学计算，将adata矩阵X中的每个元素除以其对应观测值中的”size_factors”值。第三步是在adata中添加一层，并将处理后的矩阵储存在其中，该层名为”scran_normalization”。在处理后的矩阵中，使用sc.pp.log1p对每个元素取对数并加上1，生成一个稀疏矩阵。</p>
</blockquote>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">fig, axes = plt.subplots(<span class="number">1</span>, <span class="number">2</span>, figsize=(<span class="number">10</span>, <span class="number">5</span>))</span><br><span class="line">p1 = sns.histplot(adata.obs[<span class="string">&quot;total_counts&quot;</span>], bins=<span class="number">100</span>, kde=<span class="literal">False</span>, ax=axes[<span class="number">0</span>])</span><br><span class="line">axes[<span class="number">0</span>].set_title(<span class="string">&quot;Total counts&quot;</span>)</span><br><span class="line">p2 = sns.histplot(</span><br><span class="line">    adata.layers[<span class="string">&quot;scran_normalization&quot;</span>].<span class="built_in">sum</span>(<span class="number">1</span>), bins=<span class="number">100</span>, kde=<span class="literal">False</span>, ax=axes[<span class="number">1</span>]</span><br><span class="line">)</span><br><span class="line">axes[<span class="number">1</span>].set_title(<span class="string">&quot;log1p with Scran estimated size factors&quot;</span>)</span><br><span class="line">plt.show()</span><br></pre></td></tr></table></figure>
<blockquote>
<p>画图代码同上</p>
</blockquote>
<p><img src="https://typorapicz.oss-cn-hangzhou.aliyuncs.com/img/1681887894896-71676895-64bb-4aaf-ae98-0e5eaad42474.png" alt="image.png"></p>
<h2 id="7-3-分析皮尔逊残差"><a href="#7-3-分析皮尔逊残差" class="headerlink" title="7.3. 分析皮尔逊残差"></a>7.3. 分析皮尔逊残差</h2><p>在本章中，我们介绍的第三种归一化技术是Pearson残差的分析逼近法。这种归一化技术的动机是观察到scRNA-seq数据中细胞间变异性可能会受到生物学异质性和技术效应的干扰。该方法利用“负二项式回归的正则化Pearson残差”来计算数据中的技术噪声模型。它在广义线性模型中明确地增加了计数深度作为共变量。[Germain等人，2020]在不同归一化技术的独立比较中表明，该方法消除了采样效应的影响，同时保留了数据集中的细胞异质性。值得注意的是，分析Pearson残差不需要像伪计数增加或对数转换这样的下游启发式步骤。该方法的输出是可以是正值也可以是负值的归一化值。对于细胞和基因而言，负的残差表示观察到的计数比基因的平均表达和细胞测序深度所预期的要少。正残差表示计数更多。分析Pearon残差已经在scanpy中实现，并且可以直接在原始计数矩阵上计算。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">analytic_pearson = sc.experimental.pp.normalize_pearson_residuals(adata, inplace=<span class="literal">False</span>)</span><br><span class="line">adata.layers[<span class="string">&quot;analytic_pearson_residuals&quot;</span>] = csr_matrix(analytic_pearson[<span class="string">&quot;X&quot;</span>])</span><br></pre></td></tr></table></figure>
<blockquote>
<p>这段代码用于对adata中的基因表达矩阵进行归一化处理，具体步骤如下：</p>
<ol>
<li>使用Scanpy中的<code>sc.experimental.pp.normalize_pearson_residuals</code>函数计算pearson残差，并进行归一化处理，返回值为一个Anndata对象，其中包含了处理后的归一化pearson残差矩阵。 </li>
<li>将处理后的pearson残差矩阵存储在adata的layers中，命名为”analytic_pearson_residuals”。</li>
</ol>
<p>其中，参数inplace&#x3D;False表示不改变原始的adata对象，而是返回一个新的Anndata对象。csr_matrix函数则将矩阵转化为压缩稀疏行格式，以便在内存占用上更加高效。</p>
</blockquote>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">fig, axes = plt.subplots(<span class="number">1</span>, <span class="number">2</span>, figsize=(<span class="number">10</span>, <span class="number">5</span>))</span><br><span class="line">p1 = sns.histplot(adata.obs[<span class="string">&quot;total_counts&quot;</span>], bins=<span class="number">100</span>, kde=<span class="literal">False</span>, ax=axes[<span class="number">0</span>])</span><br><span class="line">axes[<span class="number">0</span>].set_title(<span class="string">&quot;Total counts&quot;</span>)</span><br><span class="line">p2 = sns.histplot(</span><br><span class="line">    adata.layers[<span class="string">&quot;analytic_pearson_residuals&quot;</span>].<span class="built_in">sum</span>(<span class="number">1</span>), bins=<span class="number">100</span>, kde=<span class="literal">False</span>, ax=axes[<span class="number">1</span>]</span><br><span class="line">)</span><br><span class="line">axes[<span class="number">1</span>].set_title(<span class="string">&quot;Analytic Pearson residuals&quot;</span>)</span><br><span class="line">plt.show()</span><br></pre></td></tr></table></figure>
<p><img src="https://typorapicz.oss-cn-hangzhou.aliyuncs.com/img/1681888220221-c281bd7f-dfd9-494c-b33c-78c6d58e2781.png" alt="image.png"><br>我们对数据集应用了不同的归一化技术，并将它们保存为独立的层到我们的anndata对象中。根据下游分析任务的需要，使用不同归一化的层并评估结果可能更有利。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">adata.write(<span class="string">&quot;s4d8_normalization.h5ad&quot;</span>)</span><br></pre></td></tr></table></figure>
</article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta">Author: </span><span class="post-copyright-info"><a href="http://example.com">John Doe</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta">Link: </span><span class="post-copyright-info"><a href="http://example.com/2023/04/20/7.Normalization/">http://example.com/2023/04/20/7.Normalization/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta">Copyright Notice: </span><span class="post-copyright-info">All articles in this blog are licensed under <a target="_blank" rel="noopener" href="https://creativecommons.org/licenses/by-nc-sa/4.0/">CC BY-NC-SA 4.0</a> unless stating additionally.</span></div></div><div class="tag_share"><div class="post-meta__tag-list"></div><div class="post_share"><div class="social-share" data-image="https://i.loli.net/2021/02/24/5O1day2nriDzjSu.png" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2023/04/21/8.%20Feature%20selection/" title=""><div class="cover" style="background: var(--default-bg-color)"></div><div class="pagination-info"><div class="label">Previous Post</div><div class="prev_info"></div></div></a></div><div class="next-post pull-right"><a href="/2023/04/14/6.%20Quality%20Control/" title=""><div class="cover" style="background: var(--default-bg-color)"></div><div class="pagination-info"><div class="label">Next Post</div><div class="next_info"></div></div></a></div></nav></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="is-center"><div class="avatar-img"><img src="https://i.loli.net/2021/02/24/5O1day2nriDzjSu.png" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="author-info__name">John Doe</div><div class="author-info__description"></div></div><div class="card-info-data site-data is-center"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">5</div></a><a href="/tags/"><div class="headline">Tags</div><div class="length-num">0</div></a><a href="/categories/"><div class="headline">Categories</div><div class="length-num">0</div></a></div><a id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/xxxxxx"><i class="fab fa-github"></i><span>Follow Me</span></a></div><div class="card-widget card-announcement"><div class="item-headline"><i class="fas fa-bullhorn fa-shake"></i><span>Announcement</span></div><div class="announcement_content">This is my Blog</div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>Catalog</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-2"><a class="toc-link" href="#7-1-%E7%B4%A2%E5%BC%95"><span class="toc-number">1.</span> <span class="toc-text">7.1. 索引</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#7-3-%E5%88%86%E6%9E%90%E7%9A%AE%E5%B0%94%E9%80%8A%E6%AE%8B%E5%B7%AE"><span class="toc-number">2.</span> <span class="toc-text">7.3. 分析皮尔逊残差</span></a></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>Recent Post</span></div><div class="aside-list"><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2023/04/28/hello-world/" title="Hello World">Hello World</a><time datetime="2023-04-28T01:04:07.174Z" title="Created 2023-04-28 09:04:07">2023-04-28</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2023/04/22/9_Dimensionality%20Reduction/" title="No title">No title</a><time datetime="2023-04-22T13:01:20.122Z" title="Created 2023-04-22 21:01:20">2023-04-22</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2023/04/21/8.%20Feature%20selection/" title="No title">No title</a><time datetime="2023-04-21T14:35:06.487Z" title="Created 2023-04-21 22:35:06">2023-04-21</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2023/04/20/7.Normalization/" title="No title">No title</a><time datetime="2023-04-20T14:28:51.386Z" title="Created 2023-04-20 22:28:51">2023-04-20</time></div></div><div class="aside-list-item no-cover"><div class="content"><a class="title" href="/2023/04/14/6.%20Quality%20Control/" title="No title">No title</a><time datetime="2023-04-14T15:43:20.453Z" title="Created 2023-04-14 23:43:20">2023-04-14</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2020 - 2023 By John Doe</div><div class="framework-info"><span>Framework </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo</a><span class="footer-separator">|</span><span>Theme </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly</a></div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="Read Mode"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="Switch Between Light And Dark Mode"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="Toggle between single-column and double-column"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside_config" type="button" title="Setting"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="Table Of Contents"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="Back To Top"><span class="scroll-percent"></span><i class="fas fa-arrow-up"></i></button></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><script src="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox/fancybox.umd.min.js"></script><div class="js-pjax"></div><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>